portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -20,92 +20,218 @@
+     *.        direction (next or previous : string); default: next
+     *.        commentsonpage (int) (currently); default: maxcomments
+     * if parameters are invalid - redirects to '/'
+     * returns json of Commeend object
+     * CommentsSend.comments consists of maxcomments (or less) comments, which timestamp >
+     * lastTimestamp if direction = next, or <= lastTimestamp (- maxcomments comments) if direction = previous
+     */
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        final int MAX_COMMENTS_NUMBER = 1000;
 
-  /* expects maxcomments parameter (int)
-   * returns json of comments
-   * the number of comments equals to maxcomments (or less if maxcomments is
-   * greater than total number of comments in database)
-   */
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    final int MAX_COMMENTS_NUMBER = 1000;
-    // try to get maxComments parameter from request and check if it is valid
-    // if not valid - set to MAX_COMMENTS_NUMBER, i.e. return practically all comments
-    int maxNumberOfComments;
-    try {
-        maxNumberOfComments = Integer.parseInt(request.getParameter("maxcomments"));
-    } catch (NumberFormatException e) {
-        maxNumberOfComments = MAX_COMMENTS_NUMBER;
-    }
-    if (maxNumberOfComments < 0 || maxNumberOfComments > MAX_COMMENTS_NUMBER) {
-        maxNumberOfComments = MAX_COMMENTS_NUMBER;
-    }
+        // get maxComments parameter
+        String maxNumberOfCommentsString = request.getParameter("maxcomments");
+        int maxNumberOfComments;
+        if (maxNumberOfCommentsString == null) {
+            maxNumberOfComments = MAX_COMMENTS_NUMBER;
+        } else {
+            try {
+                maxNumberOfComments = Integer.parseInt(maxNumberOfCommentsString);
+            } catch (NumberFormatException e) {
+                response.sendRedirect("/");
+                return;
+            }
+        }
+        if (maxNumberOfComments < 0 || maxNumberOfComments > MAX_COMMENTS_NUMBER) {
+            response.sendRedirect("/");
+            return;
+        }
+
+        // get lastTimestamp parameter
+        String lastTimestampString = request.getParameter("timestamp");
+        long lastTimestamp;
+        if (lastTimestampString == null) {
+            lastTimestamp = System.currentTimeMillis() + 10;
+        } else {
+            try {
+                lastTimestamp = Long.parseLong(lastTimestampString);
+            } catch (NumberFormatException e) {
+                response.sendRedirect("/");
+                return;
+            }
+        }
+
+        // get direction parameter
+        String direction = request.getParameter("direction");
+        if (direction == null) {
+            direction = "next";
+        }
+        if (!direction.equals("next") && !direction.equals("previous")) {
```

Comment:
This seems like a bit odd way to compare strings. Doesn't '===' work?

Reply:
In Java '==' compares references of strings, not values, so I needed to use .equals() instead

Reply:
Got it, thanks! I learn something new every day, awesome :)